Coop-Settle channel-state logic

ezdac opened the pull request:
### Implement the Coop-Settle logic built on-top of the Withdraw protocol.
This PR implements the logic as specified in [this](https://github.com/raiden-network/spec/issues/342) proposal.
It leverages the internal locking of channels with the `NettingChannelEndState.withdraws_pending` checks, that are
already in place for the withdraws. 
The most important additional verification for coop-settle is:
- we don't accept any coop-settle related withdraws when there are pending payments (locks) for us or the partner
- we don't emit withdraw-specific events (especially contract calls) from the state-transition for coop-settle related withdraw messages.

### This PR lacks:
- Changes to the withdraw-message schema
- new API endpoint or integration in existing API endpoint that would trigger a `ActionInitCoopSettle`
- processing of the `ContractSendCoopSettle` event
- thus no blockchain interaction is implemented
- Integration tests (since the full coop-settle lifecycle is not implemented)

ezdac committed:
Add logic to implement withdraw-based coop-settle protocol

ezdac committed:
Add state-machine handling of action-coop-settle

Before, the logic to handle the `ActionChannelCoopSettle` was there,
but the channels state machine would not process the event.
Now, the dynamic type-checking of incoming events will forward
the action to the appropriate handler-function.

ezdac committed:
Add state-machine unit tests for coop-settle

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+from raiden.utils.signer import LocalSigner
+
+
+def make_receive_coop_settle_withdraw_confirmation(
```
I don't think there's need to make this public. If it's used just within this one test module, please make it private.

ezdac replied:
Nice, today I learned that `_methods` will not get imported if you do the `from foo import *`
Changed.

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+    return withdraw_confirmation
+
+
+def make_receive_coop_settle_withdraw_request(
```
Similarly here.

ezdac replied:
Done

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+    return withdraw_request
+
+
+def assert_coop_settle_state(
```
And here.

ezdac replied:
Done

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -2787,6 +2786,14 @@ def state_transition(
             block_number=block_number,
             block_hash=block_hash,
         )
+    elif type(state_change) == ActionChannelCoopSettle:
```
Shouldn't we rather use `is` for such type comparisons?

ezdac replied:
That would probably be better.
But since those type of checks are all over the place, I would rather not have this refactored in this PR, and even less so only checking with `is` for the new code path.

ulope replied:
This would be a good candidate to also refactor with `functools.singledispatch` (c.f. #7252) (in another PR)

istankovic commented on raiden/tests/unit/channel_state/test_channel_deposit.py:
```diff
@@ -0,0 +1,160 @@
+# pylint: disable=too-many-locals,too-many-statements,too-many-lines
```
I like the split! :+1:

fredo commented:
> new API endpoint or integration in existing API endpoint that would trigger a ActionInitCoopSettle

Did we agree on having a specific endpoint for it? Or couldn't we optimisticly try to coop settle and if not possible do the old settlement process?

ezdac commented:
> Did we agree on having a specific endpoint for it? Or couldn't we optimisticly try to coop settle and if not possible do the old settlement process?

That's the plan - reuse the channel-close endpoint: there try a coop-settle first and if that expires go with the traditional close/settle cycle.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
+    sender_transferred_amount = 0
+    receiver_transferred_amount = 0
+
+    if sender.balance_proof:
```
Please use explicit checks for `None`.

istankovic replied:
Done.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
+    if sender.balance_proof:
+        sender_transferred_amount = sender.balance_proof.transferred_amount
+
+    if receiver.balance_proof:
```
Please use explicit checks for `None`.

istankovic replied:
Done.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
+
+    return WithdrawAmount(
+        sender.contract_balance - sender_transferred_amount + receiver_transferred_amount
+    )
```
I would like to see the changes to `transfer/channel.py` moved into a separate commit, since this does not seem to be a part of the straightforward test-split topic.

istankovic replied:
Done.

ezdac committed:
Split tests for ChannelState

The single file testing the state-machine for
the NettingChannelState and NettingChannelEndstate
was very large and was testing a lot of different
state-changes.

In this commit, the file is split up in a separate module,
roughly grouped by the type of state-changes that are
tested

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1152,6 +1180,16 @@ def get_balance(sender: NettingChannelEndState, receiver: NettingChannelEndState
+) -> WithdrawAmount:
+    """Calculates the maximum "total_withdraw_amount" for a channel.
+    This will leave the channel without funds, when
+
```
Incomplete comment.

istankovic replied:
Fixed in a later commit.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1743,7 +1780,14 @@ def send_expired_withdraws(
+                coop_settle.total_withdraw_participant == withdraw_state.total_withdraw
+                and coop_settle.expiration == withdraw_state.expiration
+            ):
+                # We also declare the coop-settle as expired, when we initated it
```
initated -> initiated

karlb replied:
And I think it should be `if` instead of `when`.

istankovic replied:
Done.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1743,7 +1780,14 @@ def send_expired_withdraws(
+        if coop_settle is not None:
+            if (
+                coop_settle.total_withdraw_participant == withdraw_state.total_withdraw
+                and coop_settle.expiration == withdraw_state.expiration
```
Is the equality operator appropriate here when checking expiration?

karlb replied:
Since we don't allow a coop settle and withdraws at the same time, this whole if statement should always be true, shouldn't it?

istankovic replied:
Hmm... not sure about that, but looking at it the other way, I would say it seems ok to reset the `coop_settle` if _any_ withdraw expired. So I guess it could be simplified.

However, I don't like doing channel state changes here, this seems rather like a function that is supposed to return events, not cause side-effects like that. @ulope Any ideas for a better place?

ulope replied:
All these functions in here are in the end (through various layers) called from `raiden.transfer.channel.state_transition`. 
Also if you check the other functions, they all do direct state manipulation. So this is fine.

> Is the equality operator appropriate here when checking expiration?

I assume this is intended to check that the expiring withdraw is the one that belongs to the coop settle and not some other unrelated withdraw (which as @karlb said shouldn't be possible).

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1868,6 +1912,54 @@ def handle_action_close(
+    pseudo_random_generator: random.Random,
+    block_number: BlockNumber,
+) -> TransitionResult[NettingChannelState]:
+    events: List[Event] = list()
```
We should rather use a literal here: `[]`.

istankovic replied:
Done.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1868,6 +1912,54 @@ def handle_action_close(
+    )
+    partner_max_total_withdraw = get_max_withdraw_amount(
+        channel_state.partner_state, channel_state.our_state
+    )
```
These only need to be computed if `valid_coop_settle` below is `True`.

karlb replied:
But to calculate `valid_coop_settle`we need `our_max_total_withdraw`. Your statement is still true about `partner_max_total_withdraw`, but this way both of the assignments are close together.

istankovic replied:
Ah, right.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
+
+    our_initiated_coop_settle = channel_state.our_state.initiated_coop_settle
+
+    if our_initiated_coop_settle is not None or withdraw_request.coop_settle is True:
```
We should avoid checking booleans with `is`. Just use plain `or withdraw_request.coop_settle:`

istankovic replied:
Done.

karlb commented on raiden/transfer/channel.py:
```diff
@@ -1743,7 +1780,14 @@ def send_expired_withdraws(
             break
 
         nonce = get_next_nonce(channel_state.our_state)
-        channel_state.our_state.nonce = nonce
```
Why don't we need this line anymore?

ulope replied:
Good question. 
[This is the commit where it was introduced](https://github.com/raiden-network/raiden/commit/17e5ae45a147990e45ae471065ce8beb4e596a70)

istankovic replied:
I brought back the line, it's my impression that we need to update the nonce if we're to emit events that use it.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
+        if our_initiated_coop_settle is not None:
+            # There is a coop-settle inplace that we initiated
+            # and partner is communicating their total withdraw with us
+            if not our_initiated_coop_settle.expiration == withdraw_request.expiration:
```
We should use `!=` for comparison here.

istankovic replied:
Done.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
+                total_withdraw_participant=withdraw_request.total_withdraw,
+                total_withdraw_partner=our_max_total_withdraw,
+                expiration=withdraw_request.expiration,
+            )
```
Should we have a sanity check for `withdraw_request.expiration`? What happens if the partner sends us a request with unreasonably long expiration?

ulope replied:
There seems to be little incentive for the peer to do that.
Choosing a very long expiration locks funds for this period. If the peer is the one requesting the coop settle it stands to reason that they are the one with balance to gain.

Also in the worst case either party can always claim on-chain.

Some previous discussions: 
- https://github.com/raiden-network/raiden/issues/864
- https://github.com/raiden-network/raiden/pull/4289

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
+        partner_initiated_coop_settle = channel_state.partner_state.initiated_coop_settle
+        if our_initiated_coop_settle is not None:
+            # There is a coop settle inplace  that we initiated
+            assert not partner_initiated_coop_settle, "Only one party can initiate a coop settle"
```
We should check explicitly against `None`.

istankovic replied:
Done.

istankovic commented on raiden/transfer/channel.py:
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
+                    expiration=withdraw_confirmation.expiration,
+                    triggered_by_block_hash=block_hash,
+                )
+                events.append(withdraw_on_chain)
```
It would be nice to keep the old comment on this block.

istankovic replied:
Done.

istankovic commented on raiden/transfer/events.py:
```diff
@@ -51,6 +51,7 @@ class SendWithdrawRequest(SendMessageEvent):
     participant: Address
     expiration: BlockExpiration
     nonce: Nonce
+    coop_settle: Optional[bool] = False
```
Why is this optional?

karlb replied:
Because not all withdraws are coop settlements. There is still the "normal" withdraw.

ulope replied:
But what's the third state? 
`Optional[bool]` is tri-state (`True | False | None`).

I don't see `None` being used and also can't think what that would mean in this context.

istankovic replied:
I made it non-optional.

istankovic commented on raiden/transfer/events.py:
```diff
@@ -124,6 +125,40 @@ def __repr__(self) -> str:
+    our_total_withdraw: WithdrawAmount
+    partner_total_withdraw: WithdrawAmount
+    expiration: BlockExpiration
+    # This are the partner's signatures of the two withdraws.
```
This -> These

karlb commented on raiden/transfer/channel.py:
```diff
@@ -914,32 +918,58 @@ def is_valid_unlock(
+    coop_settle: ActionChannelCoopSettle,  # pylint: disable=unused-argument
+    total_withdraw: WithdrawAmount,
+) -> SuccessOrError:
+    result = is_valid_total_withdraw(channel_state, total_withdraw)
```
This will fail if we don't have anything to withdraw in our channel. It is not very useful to settle a channel if you have no balance left, but it might still be nice to allow it.

It is also not obvious why we check the total_withdraw for one participant but not the other.

ulope replied:
To me it looks like `is_valid_total_withdraw` checks both sides.

istankovic commented on raiden/transfer/events.py:
```diff
@@ -124,6 +125,40 @@ def __repr__(self) -> str:
+            f"{self.__class__.__name__}< "
+            f"canonical_identifier: {self.canonical_identifier} "
+            f"our_total_withdraw: {self.our_total_withdraw} "
+            f"parner_total_withdraw: {self.our_total_withdraw} "
```
Typo, should be
```
             f"partner_total_withdraw: {self.partner_total_withdraw} "
```

istankovic replied:
Ah, fixed in a later commit.

istankovic commented on raiden/transfer/events.py:
```diff
@@ -124,6 +125,40 @@ def __repr__(self) -> str:
+            f"parner_total_withdraw: {self.our_total_withdraw} "
+            f"expiration: {self.expiration} "
+            f"partner_signature_withdraw: {to_hex(self.signature_partner_withdraw)} "
+            f"partner_signature_our_withdraw: {to_hex(self.signature_our_withdraw)} "
```
I'd prefer to use field names as they are, instead of renaming them for the `__repr__` output.

istankovic replied:
Changed to original field names.

karlb commented on raiden/transfer/state.py:
```diff
@@ -304,6 +305,15 @@ class PendingWithdrawState:
 
+@dataclass
+class CoopSettleState:
+    total_withdraw_participant: WithdrawAmount
+    total_withdraw_partner: WithdrawAmount
```
The usual naming pair is `our` (instead of `participant`) and `partner`.

karlb replied:
I see now that these are not "our" and "our partner" but rather the participant initiating the coop settle and the participant's partner. If we don't find better names, we should at least add a short comment.

istankovic replied:
Maybe... `total_withdraw_initiator`?

istankovic replied:
Went with the initiator and left the other as partner.

karlb commented on raiden/transfer/state_change.py:
```diff
@@ -116,6 +116,18 @@ def channel_identifier(self) -> ChannelID:
 
+@dataclass(frozen=True)
+class ActionChannelCoopSettle(StateChange):
+    """Withdraw funds from channel."""
```
This comment is copy-pasted from `ActionChannelWithdraw` and needs updating.

istankovic replied:
Done.

Commit:
raiden: transfer: explicitly check for None

Commit:
raiden: transfer: fix a comment

Commit:
raiden: transfer: use a list literal

And drop the unneded type annotation.

Commit:
raiden: transfer: don't check booleans with is

Commit:
raiden: transfer: use `a != b` rather than `not a == b`

Commit:
raiden: transfer: use explicit check for None

Commit:
raiden: transfer: resurrect the old comment

It's still useful.

Commit:
raiden: transfer: fix typo

Commit:
raiden: transfer: don't rename fields in __repr__ output

Commit:
raiden: transfer: rename total_withdraw_participant to total_withdraw_initiator

Commit:
raiden: transfer: improve comment for ActionChannelCoopSettle

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+):
+    packed = pack_withdraw(
+        canonical_identifier=channel_state.canonical_identifier,
+        # pylint: disable=no-member
```
This seems unnecessary, at least with my version of pylint.

istankovic replied:
Removed this and the rest like it.

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+# pylint: disable=too-many-locals,too-many-statements,too-many-lines
```
This seems unnecessary, at least with my version of pylint.

istankovic replied:
Removed.

Commit:
tests: remove unnecessary pylint directives

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+    )
+    # The partner still has full balance in his state.
+    # He could therefore initiate a transfer in our direction now and we would accept it
+    # The then could try to continue the coop-settle with old balances.
```
Typo: "The then"

istankovic replied:
Fixed.

Commit:
tests: fix typo

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+        coop_settle_expiration=expiration,
+        initiated_coop_settle=False,
+    )
+
```
The comments here need to be swapped -- "our view" should be where "partner's view" is and vice versa.

istankovic replied:
Done.

Commit:
tests: fix comments

istankovic commented on raiden/tests/unit/channel_state/test_coop_settle.py:
```diff
@@ -0,0 +1,448 @@
+        our_initial_model=our_model1,
+        our_current_nonce=1,
+        coop_settle_expiration=expiration,
+        initiated_coop_settle=True,
```
This is not clear to me. We're setting `initiated_coop_settle` to `True`, even though it was the partner who sent the coop settle withdraw request. Could this be because we set the `CoopSettleState` manually? I'm confused.

ulope commented on raiden/transfer/channel.py:
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
+    or pending as offchain-withdraw.
 
     """
-    return WithdrawAmount(sender.total_withdraw + get_balance(sender, receiver))
```
This doesn't seem right. Why would we not need to include the previous off-chain withdraws in the calculation (which is what the `get_balance()` function does)?

ulope replied:
Ok, actually this is correct since coop settle requires no pending withdraws.
However I still think it would be nicer to use `get_balance` instead of ad-hoc reimplementing half of it here.

istankovic replied:
Done.

Commit:
raiden: transfer: don't reimplement most of get_balance

Commit:
raiden: fix rebase fallout

Commit:
raiden: transfer: bring back the update of nonce in our state

Commit:
raiden: transfer: make coop_settle non-optional
